### PR TITLE
refactor(*): make errors consistent

### PR DESCRIPTION
**Pull request by sidvishnoi**
Closes https://github.com/w3c/respec/issues/3035

Adds a `RespecError extends Error` class to make all errors consistent. `showError` and `showWarning` are improved versions of `showInlineError` and `showInlineWarning`.

**Edit:** following should help in review:
- Most relevant changes are in: `core/utils`, `core/ui` and `core/pubsubhub`.
- Then `respecDocWriter` adapts to new error format, while being backward compatible with old string-only format.
- Everything else then is replacing pub(error, msg) and showInlineError with showError (and showWarning), so viewing any one of other files should be enough.
- Side-by-side diff would be more helpful.

### CODE REVIEW

**Comment by sidvishnoi** [comment]
## Demo (first run):

In console:
image

In UI:
image

File: src/core/utils.js [hunk 1]
```diff
@@ -949,3 +950,54 @@ export class CaseInsensitiveMap extends Map {
...
   }
 }
+
+export class Err extends Error {
+  /**
+   * @param {string} message
+   * @param {string} plugin Name of plugin that caused the error.
+   * @param {object} [options]
+   * @param {boolean} [options.isWarning]
+   * @param {string} [options.title]
```
Comment (marcoscaceres) [inline src/core/utils.js:960]: "title" is Kida weird... 

```suggestion
   * @param {string} [options.message]
```

Comment (sidvishnoi) [reply src/core/utils.js:960]: `title` is what gets applied to offending elements. similar to what we had in `showInlineError`. We already have message as a required constructor parameter.

File: src/core/utils.js [hunk 2]
```diff
@@ -949,3 +898,54 @@ export class CaseInsensitiveMap extends Map {
     return super.delete(key.toLowerCase());
   }
 }
+
+export class Err extends Error {
```
Comment (sidvishnoi) [inline src/core/utils.js:902]: Consider other names: 
- `ReSpecError`: too long
- `RsError`?

File: src/core/issues-notes.js [hunk 3]
```diff
@@ -154,7 +154,8 @@ function handleIssues(ins, ghIssues, conf) {
           title.classList.add("issue-number");
           ghIssue = ghIssues.get(dataNum);
           if (!ghIssue) {
-            pub("warning", `Failed to fetch issue number ${dataNum}`);
```
Comment (sidvishnoi) [inline src/core/issues-notes.js:157]: Yay!! A bug caught :partying_face:

File: src/core/pubsubhub.js [hunk 4]
```diff
@@ -75,11 +75,12 @@ export function unsub({ topic, cb }) {
 }
 
 sub("error", err => {
-  console.error(err, err.stack);
+  console.error(err, Object.fromEntries(Object.entries(err)));
```
Comment (sidvishnoi) [inline src/core/pubsubhub.js:78]: This is similar to doing a `console.error` followed by a `console.dir`, but within the same `console.error`.

File: src/core/utils.js (hunk 1 above)
Comment (marcoscaceres) [reply src/core/utils.js:900]: ok, makes sense... maybe worth a comment so I remember next time :)

File: src/core/utils.js (hunk 2 above)
Comment (marcoscaceres) [reply src/core/utils.js]: RsError would be better IMO

File: src/core/utils.js [hunk 5]
```diff
@@ -949,3 +890,56 @@ export class CaseInsensitiveMap extends Map {
...
+  }
+
+  toHTML() {
+    const plugin = this.plugin ? `(${this.plugin}): ` : "";
+    const hint = this.hint ? ` ${this.hint}.` : "";
+    const elements = Array.isArray(this.elements)
+      ? ` Occurred at: ${joinAnd(this.elements.map(generateMarkdownLink))}.`
+      : "";
+    const details = this.details
+      ? `\n\n<details>\n${this.details}\n</details>\n`
```
Comment (marcoscaceres) [inline src/core/utils.js:931]: Any reason why we don't html`<details>? here?

Comment (marcoscaceres) [reply src/core/utils.js:931]: ah, never mind... we markdown it below

File: src/core/webidl.js [hunk 6]
```diff
@@ -184,7 +180,7 @@ function defineIdlName(escaped, data, parent) {
     const styledName = data.type === "operation" ? `${name}()` : name;
     const ofParent = parentName ? ` \`${parentName}\`'s` : "";
     const msg = `Missing \`<dfn>\` for${ofParent} \`${styledName}\` ${data.type}. [More info](https://github.com/w3c/respec/wiki/WebIDL-thing-is-not-defined).`;
-    showInlineWarning(unlinkedAnchor, msg, "");
+    pub("warn", new RsError(msg, name, { elements: [unlinkedAnchor] }));
```
Comment (marcoscaceres) [inline src/core/webidl.js:183]: yay!

File: src/core/anchor-expander.js [hunk 7]
```diff
@@ -45,7 +47,8 @@ export function run() {
       default: {
         a.textContent = a.getAttribute("href");
         const msg = "ReSpec doesn't support expanding this kind of reference.";
-        showInlineError(a, msg, `Can't expand "#${id}".`);
+        const title = `Can't expand "#${id}".`;
+        pub("error", new RsError(msg, name, { title }));
```
Comment (marcoscaceres) [inline src/core/anchor-expander.js:51]: can we pass the `a`?

Comment (sidvishnoi) [reply src/core/anchor-expander.js:51]: Can't believe I missed this one after 3 self-reviews!

File: src/core/webidl.js (hunk 6 above)
Comment (sidvishnoi) [reply src/core/webidl.js:183]: There is a subtle bug here :smile: 
`name` (plugin name) is shadowed by `name` in this `defineIdlName` function

**Comment by sidvishnoi** [comment]
This is a breaking change for respecDocWriter. Putting on hold with further breaking changes.

**Comment by sidvishnoi** [comment]
This is how information is available to the Node.js API now:
image

And to the browser UI:
image

File: tools/respecDocWriter.js [hunk 8]
```diff
@@ -271,13 +271,23 @@ async function evaluateHTML(version, timer) {
...
+        return JSON.stringify(err);
+      }
+    }, handle);
   }
 
   page.on("console", async message => {
     const args = await Promise.all(message.args().map(stringifyJSHandle));
     const msgText = message.text();
-    const text = args.filter(msg => msg !== "undefined").join(" ");
+    const text = args.filter(msg => msg !== "undefined")[0] || "";
```
Comment (sidvishnoi) [inline tools/respecDocWriter.js:290]: Note to self: this needs some attention.

File: tools/respecDocWriter.js [hunk 9]
```diff
@@ -271,13 +271,21 @@ async function evaluateHTML(version, timer) {
...
   /** @param {import('puppeteer').JSHandle<any>} handle */
   async function stringifyJSHandle(handle) {
-    return await handle.executionContext().evaluate(o => String(o), handle);
+    return await handle.executionContext().evaluate(obj => {
+      if (typeof obj === "string") {
+        // Old ReSpec versions might report errors as strings.
+        return JSON.stringify({ message: String(obj) });
+      } else {
+        // Ideally: `obj instanceof RsError` and `RsError instanceof Error`.
+        return JSON.stringify(obj);
```
Comment (sidvishnoi) [inline tools/respecDocWriter.js:280]: Note to self: A possible serialization of `RsError.elements`:
```js
if (obj.elements && Array.isArray(obj.elements)) {
  err.elements = obj.elements.map(
    /** @param {HTMLElement} el */ el => {
      el.classList.remove("respec-offending-element");
      if (el.id.includes("respec-offender")) el.id = "";
      const className = el.classList.length ? `.${el.classList}` : "";
      const id = el.id ? `#${el.id}` : "";
      const text = `[text="${el.textContent.replace(/"/g, '\\"')}"]`;
      return `${el.localName}${className}${id}${text}`;
    }
  );
}
```

**Comment by saschanaz** [comment]
This is huuuuge, could this be split somehow? 👀

**Comment by sidvishnoi** [comment]
Not sure if possible to split, as it'll leave repo in inconsistent state, but following should help in review:

- Most relevant changes are in: `core/utils`, `core/ui` and `core/pubsubhub`.
- Then `respecDocWriter` adapts to new error format, while being backward compatible with old string-only format.
- Everything else then is replacing pub(error, msg) and showInlineError with showError (and showWarning), so viewing any one of other files should be enough.

(Should've added this note first, sorry!)

**Review by marcoscaceres — APPROVED** [review]

**Pull request merged**